analyses/cell-type-ewings/scripts/singler-workflow/02-run-singler.R

```diff
@@ -0,0 +1,129 @@
+  ),
+  make_option(
+    opt_str = c("--output_annotations_file"),
+    type = "character",
+    help = "Path to TSV file to save annotations as obtained from `SingleR` for input SCE file."
+  ), 
+  make_option(
+    opt_str = c("-t", "--threads"),
+    type = "integer",
+    default = 4,
+    help = "Number of multiprocessing threads to use."
+  )
+)
+
+# Parse options
+opt <- parse_args(OptionParser(option_list = option_list))
+
+# Set up -----------------------------------------------------------------------
+
+# make sure input files exist 
+stopifnot(
+  "SCE file does not exist" = file.exists(opt$sce_file),
+  "Tumor reference file does not exist" = file.exists(opt$tumor_reference_file)
+)
+
+# create output directory if it doesn't exist
+output_dir <- dirname(opt$output_annotations_file)
+fs::dir_create(output_dir)
+
+# set up multiprocessing params
+if (opt$threads > 1) {
+  bp_param <- BiocParallel::MulticoreParam(opt$threads)
+} else {
+  bp_param <- BiocParallel::SerialParam()
+}
+
+# read in files 
+sce <- readr::read_rds(opt$sce_file)
+tumor_ref <- readr::read_rds(opt$tumor_reference_file)
+
+# Prep references --------------------------------------------------------------
+
+# grab celldex references
+blueprint_ref <- celldex::BlueprintEncodeData(ensembl = TRUE)
+hpca_ref <- celldex::HumanPrimaryCellAtlasData(ensembl = TRUE)
+
+# remove tumor cells from SCE object for annotation 
+library_id <- metadata(sce)$library_id
+sce_tumor_cells <- tumor_ref$barcodes[which(tumor_ref$library_id == library_id)]
+filtered_sce <- sce[ , !(colnames(sce) %in% sce_tumor_cells)]
+
+# remove tumor cells from library being annotated from reference 
+cells_to_keep <- tumor_ref$library_id != library_id
+filtered_tumor_ref <- tumor_ref[, cells_to_keep]
+
+# we don't need the original tumor ref anymore
+rm(tumor_ref)
+
+# Run SingleR ------------------------------------------------------------------
+
+# run SingleR on remaining cells with tumor ref and blueprint
+singler_results <- SingleR::SingleR(
+  test = filtered_sce,
+  ref = list(Blueprint = blueprint_ref,
+             HPCA = hpca_ref,
+             tumor_ref = filtered_tumor_ref),
+  labels = list(blueprint_ref$label.ont, hpca_ref$label.ont, filtered_tumor_ref$ref_tumor_label),
+  BPPARAM = bp_param
+)
+
+# get ontology labels
+cl_ont <- ontoProc::getOnto("cellOnto")
+cl_df <- data.frame(
+  annotation = cl_ont$name, # CL ID
+  ontology = names(cl_ont$name) # human readable name
+)
+
+# grab results from data frame and add human readable values for normal ont labels 
+results_df <- data.frame(
+  barcodes = rownames(singler_results),
+  ontology = singler_results$pruned.labels
+) |>
+  # replace ont labels with full human readable labels
+  dplyr::left_join(cl_df, by = c("ontology")) |>
+  # if its not found in blueprint its tumor so keep the original annotation 
+  dplyr::mutate(annotation = dplyr::if_else(is.na(annotation), ontology, annotation))
```

Comment:
An alternative to do the same thing, I think:
```suggestion
  dplyr::mutate(annotation = dplyr::coalesce(annotation, ontology))
```